adr/content-management/2023-01-06-upload-thumbnail-api-concept.md

```diff
@@ -0,0 +1,72 @@
+# 2023-01-06 - Upload Thumbnails API
+
+## Context
+We’re building an App that should optimize Thumbnails via the App. Currently, it’s not possible to upload Thumbnails via the API. All Thumbnails are processed by Shopware itself. There’s no way to upload your optimized Thumbnails via the API.
+
+However: Since it’s an App, we need to be able to do so via the API.
+
+## Decision
+### Migration
+Add a new column `isUploaded` into the media_thumbnail table. This flag lets us know if the thumbnail is uploaded or generated.
+
+### Core
+
+Add a new `updloadThumbnail` API in Shopware\Core\Content\Media\Api\MediaUploadController
+```php
+/**
+ * @Route("/api/_action/media/{mediaId}/upload-thumbnail", name="api.action.media.upload-thumbnail", methods={"POST"})
+ */
+public function uploadThumbnail(string $mediaId, Request $request, Context $context, ResponseFactoryInterface $responseFactory): Response
+{
+    // Based on mediaId, validate uploaded thumbnails has to map with the media thumbnail sizes
+
+    // call thumbnail service to upload thumbnails
+}
+```
+
+Add a new function in Shopware\Core\Content\Media\Thumbnail\ThumbnailService
+```php
+public function uploadThumbnails(string $mediaId, ThumbnailCollection $uploadedThumbnails, Context $context): void
+{
+    // we might need to check if thumbnails of the $mediaId have already been uploaded,
+    // should we have the option to skip or replace them?
```

Comment:
I would just replace it always. The deletion of a thumbnail should work just normal with the normal API CRUD.
You should check here, that on deletion the thumbnail file is also deleted

Reply:
@shyim In case we replace them. IMO, if the user wants to cancel the subscription from a plugin or an app, they will expect the thumbnails back to default, so we need an event to generate thumbnails by Shopware default for all of the media again. I don't know if it will be an issue with the performance and user experience :D

Reply:
Wa wanted to keep the default behavior whenever it's possible - Not sure if replacing the default ones is a good idea? 
If you want to generate your own Thumbnails - You want to create them on top of the existing and don't delete anything existing.

Also we wanted to consider the case as @hungmac-sw mentioned.

If a user uninstalls the App - The user want's to get back to the default - Which would not be possible if we're overriding aKa deleting existing data/thumbnails.

Even though: The Thumbnails are generated from the original image ( which we're not touching ).
So i guess to get back to the default: The user could just generate the Thumbnails again to have the same ones as before @hungmac-sw